### DISCUSSION TITLE
[FEATURE] Implement missing AWS Greengrass integration that's present in the AWS SDK

### DISCUSSION THREAD

**Discussion by pilartomas** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#961] [Ideas]
**Is your feature request related to a problem? Please describe.**
Hello, I'm trying to properly integrate Kinesis Video Streams Producer SDK (used to build GStreamer sink) as a component into my AWS Greengrass core device. However, I've found that KVS SDK is missing a Greengrass integration that's normally present in the AWS SDK.

**Describe the solution you'd like**
Implement the feature from the AWS SDK. Following is my estimation of what the AWS SDK does.

When no credentials are given to the SDK, as a fallback, look for `AWS_CONTAINER_CREDENTIALS_FULL_URI` environment variable. If it exists, use the local [Token Exchange service](https://docs.aws.amazon.com/greengrass/v2/developerguide/token-exchange-service-component.html) to get temporary credentials for the role.

**Describe alternatives you've considered**
KVS SDK is currently giving you the option to specify credentials (X. 509 cert, private key, role alias, root CA cert) directly and handles the temporary tokens on it's own. However, that is not ideal and bypasses proper component composition in Greengrass.

**Additional context**
[Greengrass interacting with AWS services](https://docs.aws.amazon.com/greengrass/v2/developerguide/interact-with-aws-services.html)
[Greengrass environment variables](https://docs.aws.amazon.com/greengrass/v2/developerguide/component-environment-variables.html)

Hopefully, the info is clear enough. My apologies in case the feature has already been implemented and I am just unable find it in the codebase.

**Comment by bpawlik** [comment]
How about you use the solution presented in `awstreamer`: https://github.com/awslabs/aws-streamer/tree/master/examples/greengrass ?

**Comment by nbaillie** [comment]
I have a similar situation and the difference from the example above is that i am using the gstreamer in a container (intel/dlstreamer), so not quite the same. and would be good to have this work when passing the AWS_CONTAINER_CREDENTIALS_FULL_URI to the container.
